crates/oxc_transformer/src/es2016/exponentiation_operator.rs

```diff
@@ -0,0 +1,322 @@
+use oxc_ast::ast::*;
+use oxc_semantic::{ReferenceFlag, SymbolFlags};
+use oxc_span::SPAN;
+use oxc_syntax::operator::{AssignmentOperator, BinaryOperator};
+use oxc_traverse::TraverseCtx;
+
+use crate::context::Ctx;
+
+/// ES2016: Exponentiation Operator
+///
+/// References:
+/// * <https://babel.dev/docs/babel-plugin-transform-exponentiation-operator>
+/// * <https://github.com/babel/babel/blob/main/packages/babel-plugin-transform-exponentiation-operator>
+/// * <https://github.com/babel/babel/blob/main/packages/babel-helper-builder-binary-assignment-operator-visitor>
+pub struct ExponentiationOperator<'a> {
+    _ctx: Ctx<'a>,
+    var_declarations: std::vec::Vec<Vec<'a, VariableDeclarator<'a>>>,
+}
+
+#[derive(Debug)]
+struct Exploded<'a> {
+    reference: AssignmentTarget<'a>,
+    uid: Expression<'a>,
+}
+
+impl<'a> ExponentiationOperator<'a> {
+    pub fn new(ctx: Ctx<'a>) -> Self {
+        Self { _ctx: ctx, var_declarations: vec![] }
+    }
+
+    fn clone_identifier_reference(
+        ident: &IdentifierReference<'a>,
+        ctx: &mut TraverseCtx<'a>,
+    ) -> IdentifierReference<'a> {
+        let reference = ctx.symbols().get_reference(ident.reference_id.get().unwrap());
+        let symbol_id = reference.symbol_id();
+        let flag = reference.flag();
+        ctx.create_reference_id(ident.span, ident.name.clone(), symbol_id, *flag)
+    }
+
+    fn clone_expression(expr: &Expression<'a>, ctx: &mut TraverseCtx<'a>) -> Expression<'a> {
+        match expr {
+            Expression::Identifier(ident) => ctx
+                .ast
+                .expression_from_identifier_reference(Self::clone_identifier_reference(ident, ctx)),
+            _ => expr.clone_in(ctx.ast.allocator),
+        }
+    }
+
+    pub fn transform_statements(
+        &mut self,
+        _statements: &mut Vec<'a, Statement<'a>>,
+        ctx: &mut TraverseCtx<'a>,
+    ) {
+        self.var_declarations.push(ctx.ast.vec());
+    }
+
+    pub fn transform_statements_on_exit(
+        &mut self,
+        statements: &mut Vec<'a, Statement<'a>>,
+        ctx: &mut TraverseCtx<'a>,
+    ) {
+        if let Some(declarations) = self.var_declarations.pop() {
+            if declarations.is_empty() {
+                return;
+            }
+            let variable = ctx.ast.alloc_variable_declaration(
+                SPAN,
+                VariableDeclarationKind::Var,
+                declarations,
+                false,
+            );
+            statements.insert(0, Statement::VariableDeclaration(variable));
+        }
+    }
+
+    pub fn transform_expression(&mut self, expr: &mut Expression<'a>, ctx: &mut TraverseCtx<'a>) {
+        // left ** right
+        if let Expression::BinaryExpression(binary_expr) = expr {
+            if binary_expr.operator == BinaryOperator::Exponential {
+                let left = ctx.ast.move_expression(&mut binary_expr.left);
+                let right = ctx.ast.move_expression(&mut binary_expr.right);
```

Comment:
Thank you for not using `ast.copy` here! ❤️

Reply:
Haha, If I still use `ast.copy`, it would cause tons of semantic errors after #4790